pandas/core/arrays/floating.py

```diff
@@ -0,0 +1,641 @@
+import numbers
+from typing import TYPE_CHECKING, List, Optional, Tuple, Type, Union
+import warnings
+
+import numpy as np
+
+from pandas._libs import lib, missing as libmissing
+from pandas._typing import ArrayLike, DtypeObj
+from pandas.compat import set_function_name
+from pandas.compat.numpy import function as nv
+from pandas.util._decorators import cache_readonly
+
+from pandas.core.dtypes.cast import astype_nansafe
+from pandas.core.dtypes.common import (
+    is_bool_dtype,
+    is_datetime64_dtype,
+    is_float,
+    is_float_dtype,
+    is_integer,
+    is_integer_dtype,
+    is_list_like,
+    is_object_dtype,
+    pandas_dtype,
+)
+from pandas.core.dtypes.dtypes import register_extension_dtype
+from pandas.core.dtypes.missing import isna
+
+from pandas.core import nanops, ops
+from pandas.core.array_algos import masked_reductions
+from pandas.core.ops import invalid_comparison
+from pandas.core.ops.common import unpack_zerodim_and_defer
+from pandas.core.tools.numeric import to_numeric
+
+from .masked import BaseMaskedArray, BaseMaskedDtype
+
+if TYPE_CHECKING:
+    import pyarrow  # noqa: F401
+
+
+class _FloatingDtype(BaseMaskedDtype):
+    """
+    An ExtensionDtype to hold a single size of floating dtype.
+
+    These specific implementations are subclasses of the non-public
+    _FloatingDtype. For example we have Float32Dtype to represent float32.
+
+    The attributes name & type are set when these subclasses are created.
+    """
+
+    name: str
+    base = None
+    type: Type
+
+    def __repr__(self) -> str:
+        return f"{self.name}Dtype()"
+
+    @property
+    def _is_numeric(self) -> bool:
+        return True
+
+    @cache_readonly
+    def numpy_dtype(self) -> np.dtype:
+        """ Return an instance of our numpy dtype """
+        return np.dtype(self.type)
+
+    @cache_readonly
+    def kind(self) -> str:
+        return self.numpy_dtype.kind
+
+    @cache_readonly
+    def itemsize(self) -> int:
+        """ Return the number of bytes in this dtype """
+        return self.numpy_dtype.itemsize
+
+    @classmethod
+    def construct_array_type(cls) -> Type["FloatingArray"]:
+        """
+        Return the array type associated with this dtype.
+
+        Returns
+        -------
+        type
+        """
+        return FloatingArray
+
+    def _get_common_dtype(self, dtypes: List[DtypeObj]) -> Optional[DtypeObj]:
+        # for now only handle other floating types
+        if not all(isinstance(t, _FloatingDtype) for t in dtypes):
+            return None
+        np_dtype = np.find_common_type(
+            [t.numpy_dtype for t in dtypes], []  # type: ignore
```

Comment:
what error was this producing?

Reply:
@WillAyd This was added in https://github.com/pandas-dev/pandas/pull/33607. 
The problem is that by doing the `if not all(isinstance(t, _FloatingDtype) for t in dtypes): return None`, we know that by ending up here, we have a list of all FloatingDtype dtypes (which have a  `numpy_dtype` attribute). But mypy doesn't know this. 
A `cast(List[_FloatingDtype], dtypes)` might help?